everestrs/everestrs/src/lib.rs

```diff
@@ -0,0 +1,196 @@
+        // Wherever this shows it means "pointer to some random blob of memory that only Rust
+        // understand what it means and is of no concern to C".
+        obj: *const c_char,
+    }
+
+    struct JsonBlob {
+        data: Vec<u8>,
+    }
+
+    unsafe extern "C++" {
+        include!("everestrs_sys/everestrs_sys.hpp");
+
+        type Module;
+        fn create_module(module_id: &str, prefix: &str, conf: &str) -> UniquePtr<Module>;
+
+        /// Connects to the message broker and launches the main everest thread to push work
+        /// forward. Returns the module manifest.
+        fn initialize(self: Pin<&mut Module>) -> JsonBlob;
+
+        /// Returns the interface definition.
+        fn get_interface(self: &Module, interface_name: &str) -> JsonBlob;
+
+        /// Registers the callback of the `GenericModule` to be called and calls
+        /// `Everest::Module::signal_ready`.
+        unsafe fn signal_ready(
+            self: &Module,
+            // See comment in `CommandMeta::obj`.
+            obj: *const c_char,
+            on_ready: unsafe fn(obj: *const c_char) -> (),
+        );
+
+        /// Informs the runtime that we implement the command described in `meta` and registers the
+        /// `handle_command` method from the `GenericModule` as the handler.
+        unsafe fn provide_command(
+            self: &Module,
+            meta: CommandMeta,
+            on_ready: unsafe fn(&CommandMeta, JsonBlob) -> JsonBlob,
+        );
+    }
+}
+
+impl ffi::JsonBlob {
+    fn as_bytes(&self) -> &[u8] {
+        &self.data
+    }
+
+    fn deserialize<T: DeserializeOwned>(self) -> T {
+        // TODO(hrapp): Error handling
+        serde_json::from_slice(self.as_bytes()).unwrap()
+    }
+
+    fn from_vec(data: Vec<u8>) -> Self {
+        Self { data }
+    }
+}
+
+#[derive(FromArgs, Debug)]
+/// An everest Node.
+struct Args {
+    /// prefix of installation.
+    #[argh(option)]
+    #[allow(unused)]
+    pub prefix: PathBuf,
+
+    /// configuration yml that we are running.
+    #[argh(option)]
+    #[allow(unused)]
+    pub conf: PathBuf,
+
+    /// module name for us.
+    #[argh(option)]
+    pub module: String,
+}
+
+/// Implements the handling of commands & variables, but has no specific information about the
+/// details of the current module, i.e. it deals with JSON blobs and strings as command names. Code
+/// generation is used to build the concrete, strongly typed abstractions that are then used by
+/// final implementors.
+pub trait GenericModule: Sync {
+    /// Handler for the command `name` on `implementation_id` with the given `parameters`. The return value
+    /// will be returned as the result of the call.
+    fn handle_command(
+        &self,
+        implementation_id: &str,
+        name: &str,
+        parameters: HashMap<String, serde_json::Value>,
+    ) -> Result<serde_json::Value>;
+
+    fn on_ready(&self) {}
+}
+
+pub struct Runtime<T: GenericModule> {
```

Comment:
This is like the runtime handle, which keeps the rust module alive and tracks its link to the cxx side "module".
Concerning the naming: we used `Runtime` usually for the runtime setup, that is, where is everest located (prefix) and the logging config I think.  This runtime can then be used to spawn a new module (or in everestpy even multiple modules).  I personally like this flexibility to create/implement various modules inside one unit for running tests and things like that.
Concerning the link between both, why not having the rust module struct having it's one member variable, linking to the cxx module?

Reply:
> I personally like this flexibility to create/implement various modules inside one unit for running tests and things like that.
I like that too, but I also find the danger of misuse high: instantiating twice the same module would yield strange behaviour, but is not technically impossible. Here it is. I can see the usefulness to bring up multiple modules in the same binary for testing, but I'd favor the approach to give each of them their own `Runtime` - or `EVerest handle`.

>  Concerning the link between both, why not having the rust module struct having it's one member variable, linking to the cxx module?

I could not parse that. What is it you are suggesting?